File: web/dashboard/src/types/trace.ts
/** Matches internal/domain/trace.go */

export type StepStatus = 'success' | 'failure' | 'skipped';

export type MissionOutcome = 'success' | 'failure' | 'partial' | 'timeout';

export interface TraceStep {
  spanId: string;
  traceId: string;
  parentSpanId: string;
  name: string;
  startTime: string;
  duration: number; // milliseconds
  status: StepStatus;
  attributes: Record<string, string>;
}

export interface MissionTrace {
  traceId: string;
  missionId: string;
  agentId: string;
  goal: string;
  steps: TraceStep[];
  outcome: MissionOutcome;
  startTime: string;
  endTime?: string;
  duration: number; // milliseconds
}

export interface DriftAlert {
  agentId: string;
  metricName: string;
  observedValue: number;
  baselineValue: number;
  deviation: number;
  detectedAt: string;
}

File: internal/domain/trace.go
package domain

import "time"

// StepStatus represents the outcome of a trace step.
type StepStatus string

const (
	StepStatusSuccess StepStatus = "success"
	StepStatusFailure StepStatus = "failure"
	StepStatusSkipped StepStatus = "skipped"
)

// MissionTrace captures the full execution trace of a mission.
type MissionTrace struct {
	TraceID   string
	MissionID string
	AgentID   string
	Goal      string
	Steps     []TraceStep
	Outcome   MissionOutcome
	StartTime time.Time
	EndTime   *time.Time
	Duration  time.Duration
	ExpiresAt time.Time
}

// TraceStep represents a single span within a mission trace.
type TraceStep struct {
	SpanID       string
	TraceID      string
	ParentSpanID string
	Name         string
	StartTime    time.Time
	Duration     time.Duration
	Status       StepStatus
	Attributes   map[string]string
}

// TraceSpan is the input for adding a span to a trace.
type TraceSpan struct {
	SpanID       string
	TraceID      string
	ParentSpanID string
	Name         string
	StartTime    time.Time
	DurationMs   int64
	Status       StepStatus
	Attributes   map[string]string
}

// TraceFilter defines criteria for querying traces.
type TraceFilter struct {
	MissionID string
	AgentID   string
	Outcome   MissionOutcome
	StartTime time.Time
	EndTime   time.Time
	Limit     int
}

// DriftAlert represents a detected behavioral drift.
type DriftAlert struct {
	AgentID       string
	MetricName    string
	ObservedValue float64
	BaselineValue float64
	Deviation     float64
	DetectedAt    time.Time
}
